### ISSUE TITLE
Some filter rules suddenly stopped working

### ISSUE THREAD

**Issue by bdr99** [uBlockOrigin/uBlock-issues#3050]
### Prerequisites

- [X] I verified that this is not a filter list issue. Report any issues with filter lists or broken website functionality in the [uAssets issue tracker](https://github.com/uBlockOrigin/uAssets/issues).
- [X] This is NOT an issue with [YouTube](https://www.github.com/uBlockOrigin/uAssets/issues/20586), [Facebook](https://www.github.com/uBlockOrigin/uAssets/issues/3367) or [Twitch](https://www.github.com/uBlockOrigin/uAssets/issues/5184).
- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] The issue is not present after disabling uBO in the browser.
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I am reporting is not normal behavior.

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

I noticed that recently, certain filter rules are no longer being applied, for seemingly no reason.

For example, EasyList contains this rule which is supposed to remove ads on Reddit:
```
reddit.com,reddittorjg6rue252oqsxryoxengawnmo46qy4kyii5wtqnwfj4ooad.onion##.promotedlink:not([style^="height: 1px;"])
```

But when I navigate to a Reddit page that contains ads, such as https://old.reddit.com/r/uBlockOrigin/, I still see the "promoted" posts. If I check the logger, the rule is shown and highlighted in yellow, which I believe means it should have been applied.

![image](https://github.com/uBlockOrigin/uBlock-issues/assets/2292715/958096e8-0be0-4d2b-9fd6-2eccb52a9f2a)

I have tried the following:

| Attempted Solution | Outcome |
| ------------- | ------------- |
| Disabled all add-ons except uBO  | **No change.**  |
| Reset uBO using the "Reset to default settings" button  | **No change.** |
| Used a new unmodified browser profile  | **This does fix the issue, but I'm trying to fix it on my existing browser profile.** |
| Added the above filter rule to "My filters"  | **No change.** |
| Unchecked "Parse and enforce cosmetic filters"  | **No change.** |
| Added the rule to "My filters" AND unchecked "Parse and enforce cosmetic filters"  | **This works and the rule is applied correctly.** |
| Held shift, clicked "Purge all caches", then updated filter lists. | **No change.** |
| Uninstalled and reinstalled uBO. | **No change.** |

### A specific URL where the issue occurs.

https://old.reddit.com/r/uBlockOrigin/

### Steps to Reproduce

1. Install uBO with default settings and filter lists.
2. Navigate to https://old.reddit.com/r/uBlockOrigin/.

### Expected behavior

Promoted posts should not be visible.

### Actual behavior

Promoted posts are visible.

### uBO version

1.54.0

### Browser name and version

Firefox 121.0

### Operating System and version

Windows 11

**Comment by gwarser** [comment]
Please share your troubleshooting information. Go to Reddit, click uBO icon on toolbar, the "chat" button, "Troubleshooting Information" at the bottom, select all and copy here.

**Comment by bdr99** [comment]
@gwarser Thanks for the reply, here is my troubleshooting info:
```
uBlock Origin: 1.54.0
Firefox: 121
filterset (summary):
 network: 104575
 cosmetic: 44290
 scriptlet: 18655
 html: 1978
listset (total-discarded, last-updated):
 default:
  user-filters: 0-0, never
  ublock-filters: 37241-151, 12h.10m
  ublock-badware: 7569-11, 12h.10m
  ublock-privacy: 948-4, 12h.10m
  ublock-unbreak: 2222-1, 12h.10m
  ublock-quick-fixes: 152-0, 12h.10m
  easylist: 76499-670, 12h.10m
  easyprivacy: 33170-811, 12h.10m
  urlhaus-1: 9710-0, 12h.10m
  plowe-0: 3784-1, 12h.10m
filterset (user): [empty]
userSettings: [none]
hiddenSettings: [none]
supportStats:
 allReadyAfter: 318 ms
 maxAssetCacheWait: 131 ms
popupPanel:
 blocked: 12
 network:
  reddit.com: 9
  ad-delivery.net: 1
  redditmedia.com: 1
  redditstatic.com: 1
 extended:
  ##a.outbound[data-outbound-url]:remove-attr(data-outbound-url)
  ##+js(abort-on-property-write, STICKY_CANARY)
```

**Comment by gwarser** [comment]
Can you reapeat the logging and click the `.promotedlink...` filter to see details? Does it tell the filter is from EasyList?

Try if clicking "Purge all caches" with Shift pressed and the updating filter lists improves something.

**Comment by bdr99** [comment]
@gwarser

This is what I see if I click on the filter in the logger.

![image](https://github.com/uBlockOrigin/uBlock-issues/assets/2292715/2cf5fe19-752e-4b57-800c-8289d6b46c47)

I tried holding shift and clicking "Purge all caches" and then updating the filter lists, but this didn't help unfortunately. I also tried uninstalling uBO entirely and installing it again, but this didn't help either.

**Comment by gwarser** [comment]
If you switch the logger to DOM inspector (the `</>` button in top right) did you see the `.promotedlink...` filter blocking something?

I think it's Reddit doing some A/B testing.

**Comment by bdr99** [comment]
> If you switch the logger to DOM inspector (the `</>` button in top right) did you see the `.promotedlink...` filter blocking something?

If I click on the DOM inspector button, I actually just get a blank screen.

![image](https://github.com/uBlockOrigin/uBlock-issues/assets/2292715/aadb3a18-56c2-4370-b18e-a3154e0784cd)

> I think it's Reddit doing some A/B testing.

But the strange part is that (as mentioned in the issue description) if I add that same exact rule to "my filters", it works (as long as I also uncheck "Parse and enforce cosmetic filters").

**Comment by u-RraaLL** [comment]
Install dev build and test: https://github.com/gorhill/uBlock/releases/tag/1.54.1rc1

To install click the signed firefox file. Proceed with accepting, then right-click uBO icon > Manage extension > Update now.

**Comment by bdr99** [comment]
@u-RraaLL No change, I still get the same issue with 1.54.1rc1.

**Comment by bdr99** [comment]
Also, I checked the DOM, and I can see that the reddit ad div still has the `promotedlink` CSS class. So it definitely seems like it should still match that rule.

![image](https://github.com/uBlockOrigin/uBlock-issues/assets/2292715/1af18389-421e-4f8c-a3cd-303289c9ff9a)

**Comment by gorhill** [comment]
Maybe one invalid cosmetic filter being injected is breaking all others. Would need to find which cosmetic filter from which list.

**Comment by gorhill** [comment]
To help narrow the problematic cosmetic filter, first, try:

   @@||reddit.com^$ghide

This will tell us whether the problematic filter is generic or specific.

**Comment by bdr99** [comment]
> Maybe one invalid cosmetic filter being injected is breaking all others. Would need to find which cosmetic filter from which list.

When I test with a clean Firefox profile, I don't see this issue. Wouldn't this indicate that it isn't caused by a filter rule?

> To help narrow the problematic cosmetic filter, first, try:
> 
> ```
> @@||reddit.com^$ghide
> ```
> 
> This will tell us whether the problematic filter is generic or specific.

I added this rule to "My rules" but there was no change, I still see the ad.

> Also, looking at the ~page~ logger dev tools console when enabling dom inspector might show an selector error being reported.

It doesn't look like there is any output in the logger dev tools console when enabling the DOM inspector.

![image](https://github.com/uBlockOrigin/uBlock-issues/assets/2292715/cb8dfaef-15aa-4d51-8d41-8c1def2c4462)

**Comment by gorhill** [comment]
> Used a new unmodified browser profile
> his does fix the issue, but I'm trying to fix it on my existing browser profile.

Did you set `layout.css.has-selector.enabled` to `false` in `about:config`? I can reproduce all your symptoms when setting it to `false`. It default to `true` as of Firefox 121.

**Comment by garry-ut99** [comment]
> bdr99 : https://github.com/uBlockOrigin/uBlock-issues/issues/3050#issuecomment-1872585929 : But the strange part is that (as mentioned in the issue description) if I add that same exact rule to "my filters", it works 

Nothing strange, it's by design, unchecking this option disables all cosmetic filters in all filter lists, except "My Filters pane", that's why your filter still works, like already mentioned in :
- https://github.com/uBlockOrigin/uBlock-issues/issues/821#issuecomment-567405096
- https://github.com/uBlockOrigin/uBlock-issues/wiki/Dashboard:-Filter-lists#parse-and-enforce-cosmetic-filters, 

<hr>

> bdr99 : https://github.com/uBlockOrigin/uBlock-issues/issues/3050#issuecomment-1872585929 : (as long as I also uncheck "Parse and enforce cosmetic filters").

Why it works only after disabling cosmetic filters in all lists, is another story, maybe because like gorhill already said : maybe one of lists/filter breaking the other.

> gorhill : https://github.com/uBlockOrigin/uBlock-issues/issues/3050#issuecomment-1872605039 : Maybe one invalid cosmetic filter being injected is breaking all others. Would need to find which cosmetic filter from which list.

Similiar to the old story with : https://github.com/uBlockOrigin/uBlock-issues/issues/1751, but this time in Firefox.

**Comment by gorhill** [comment]
If `reddit.com##.promotedlink:not([style^="height: 1px;"])` work in "My filters", then add `reddit.com##.size-compact.Post:has([class*="promoted"])` and see if the issue comes back.

**Comment by bdr99** [comment]
> Did you set `layout.css.has-selector.enabled` to `false` in `about:config`? I can reproduce all your symptoms when setting it to `false`. It defaults to `true` as of Firefox 121.

Yes, I actually did previously set it to `false` in my user.js to work around a rendering bug I was experiencing in Firefox 117. I just tried setting it to `true` and it fixes my issue. So the issue only occurs with `layout.css.has-selector.enabled` set to `false`. This also made the DOM inspector start working again.

I also noticed that if I disable the "uBlock filters – Ads" list, it fixes the issue as well, even with `layout.css.has-selector.enabled` still set to false. So maybe the rule from EasyList is conflicting with a rule from this list.

> If `reddit.com##.promotedlink:not([style^="height: 1px;"])` works in "My filters", then add `reddit.com##.size-compact.Post:has([class*="promoted"])` and see if the issue comes back.

Yes, it was exactly like you predicted. After unchecking "Parse and enforce cosmetic filters", adding your first rule fixed the issue, and adding your second rule made the issue come back.

**Comment by garry-ut99** [comment]
Same here, with one exception :

> bdr99 : and adding your second rule made the issue come back.

On Firefox 115.6.0 ESR adding the second rule doesn't make the issue come back for me.

**Comment by gorhill** [comment]
Because Firefox 115 does not report supporting native `:has()`, so uBO's procedural code path is taken.

**Comment by garry-ut99** [comment]
Sorry, I don't understand your answer, since we're talking about a case with `layout.css.has-selector.enabled` set to `false` in both FF 115 and FF 121, so both of them report as not supporting native `:has()` (`if ( CSS.supports('selector(a:has(b))') )`).

>  garry-ut99 : On Firefox 115.6.0 ESR adding the second rule doesn't make the issue come back for me.

Not anymore, today adding the second rule does make the issue come back, I'm not sure whether because content of filters lists has changed, or content of reddit site has changed.

**Comment by gwarser** [comment]
@garry-ut99 not sure if this make it more clear, but the `CSS.supports` is recent https://github.com/gorhill/uBlock/commit/38bde79146be8f697693dc0eb16b5b9c168f9dd4 previously it was simple version check https://github.com/gorhill/uBlock/commit/c5724c1cce

**Comment by gorhill** [comment]
> `layout.css.has-selector.enabled` set to `false` in both FF 115 and FF 121

It's not a given that `layout.css.has-selector.enabled` to `false` will cause `CSS.supports('a:has(b)')` to return `false` now that `:has()` is officially supported.

**Comment by garry-ut99** [comment]
> gwarser : not sure if this answers your question (...)

It doesn't answer my question, because I use the newest uBO dev (1.54.1rc1).

> gorhill : It's not a given that (...)

Nor it's given, that setting it to `false` is supposed to not work anymore, if a flag doesn't works anymore, then should be removed, or be clear information somewhere, otherwise, it misleads users.

**Comment by gorhill** [comment]
> then should be removed

No point arguing that here, just open an issue at bugzilla.

**Comment by garry-ut99** [comment]
Not me who started "it's not a given" odd arguing in this place, I only replied in the same place.

**Comment by krystian3w** [comment]
> No point arguing that here, just open an issue at bugzilla.

It turns out that since release 142, this preference has been removed and its false status is ignored.


It doesn't look like anyone requested this on Bugzilla, more likely on GitHub, or exercised their right to be forgotten on Bugzilla (It is doubtful that the request to remove preferences was given the status that requires high privileges on the Bugzilla account).

So the group of affected users is limited to the Firefox 140 ESR line (From June, it will be possible to switch to version 153, and from September, line 140 will be phased out of support), with the old uBO.
